### DISCUSSION TITLE
[Question] Best way to delete an element from a store array

### DISCUSSION THREAD

**Discussion by i9kolp** [solidjs/solid#1401] [Q&A] [answered]
Hi All,

I'm currently learning SolidJS, and I had a question regarding the best ways to delete an element from a store array. The following was code adapted from the todo example.

```js
import { render } from "solid-js/web";
import { For } from "solid-js";
import { createStore } from "solid-js/store";

const App = () => {
  const [todos, setTodos] = createStore([]);

  let input;
  let todoId = 0;

  const addTodo = (event) => {
    event.preventDefault();
    if (!input.value.trim()) return;
    setTodos([...todos, { id: ++todoId, text: input.value, completed: false }]);
    input.value = "";
    input.focus();
  };

  const toggleTodo = (id) => {
    setTodos(
      (todo) => todo.id === id,
      "completed",
      (completed) => !completed
    );
  };

  const deleteTodo = (id) => {
    setTodos((prevTodos) => prevTodos.filter((todo) => todo.id !== id));
  };

  return (
    <>
      <form onSubmit={addTodo}>
        <input ref={input} />
        <button type="submit">Add</button>
      </form>
      <For each={todos}>
        {(todo) => (
          <div>
            <input type="checkbox" checked={todo.completed} onChange={() => toggleTodo(todo.id)} />
            <span style={{ "text-decoration": todo.completed ? "line-through" : "none" }}>{todo.text}</span>
            <button onClick={() => deleteTodo(todo.id)}>Delete</button>
          </div>
        )}
      </For>
    </>
  );
};

render(App, document.getElementById("app"));
```

In my `deleteTodo` function, is using `filter` the best way to delete the target element? I read that "diff" the entire state is not great for performance. However, I didn't find another way from the [API docs](https://www.solidjs.com/docs/latest/api#updating-stores).

Thanks for your help!

**Comment by ryansolid** [comment, marked as answer]
Yeah filter. Or clone, splice and set I think is technically faster. But yeah it is fine. It only clones the array, and is not a deep clone. Yes it requires downstream list reconciliation but we do lists that way anyway, since untangling batched updates on array operations ends up not any more performant and it is more complicated.

**Reply by i9kolp** [reply]
Thank you @ryansolid 

SolidJS is pretty amazing BTW!
